src/reducers/existingThoughtMove.js

```diff
@@ -58,23 +61,45 @@ export default (state, { oldPath, newPath }) => {
   // if the contexts have changed, remove the value from the old contextIndex and add it to the new
   const subthoughtsOld = (state.contextIndex[contextEncodedOld] || [])
     .filter(child => !equalThoughtRanked(child, { value, rank: oldRank }))
-  const subthoughtsNew = (state.contextIndex[contextEncodedNew] || [])
-    .filter(child => !equalThoughtRanked(child, { value, rank: oldRank }))
-    .concat({
-      value,
-      rank: newRank,
-      lastUpdated: timestamp()
-    })
-
-  const recursiveUpdates = (thoughtsRanked, contextRecursive = [], accumRecursive = {}) => {
-
-    return getThoughtsRanked(thoughtsRanked, state.thoughtIndex, state.contextIndex).reduce((accum, child) => {
+  const subthoughtsNew = sameContext ?
+    (state.contextIndex[contextEncodedNew] || [])
+      .filter(child => !equalThoughtRanked(child, { value, rank: oldRank }))
+      .concat({
+        value,
+        rank: newRank,
+        lastUpdated: timestamp()
+      })
+    :
+    _.uniqBy((state.contextIndex[contextEncodedNew] || [])
+      .concat({
+        value,
+        rank: newRank,
+        lastUpdated: timestamp()
+      }),
+      "value"
+    );
+
+  const newLastRank = getNextRank(newPath, state.thoughtIndex, state.contextIndex);
+
+  const recursiveUpdates = (thoughtsRanked, contextRecursive = [], accumRecursive = {}, recursiveDepth = 0) => {
+
+    return getThoughtsRanked(thoughtsRanked, state.thoughtIndex, state.contextIndex).reduce((accum, child, i) => {
       const hashedKey = hashThought(child.value)
       const childThought = getThought(child.value, thoughtIndex)
 
       // remove and add the new context of the child
       const contextNew = newThoughts.concat(contextRecursive)
-      const childNew = addContext(removeContext(childThought, pathToContext(thoughtsRanked), child.rank), contextNew, child.rank)
+
+      // update rank of first depth of childs
+      const movedRank = recursiveDepth ? child.rank : newLastRank + i
+
+      const childNewThought = addContext(removeContext(childThought, pathToContext(thoughtsRanked), child.rank), contextNew, movedRank)
+
+      // remove duplicated context
+      const childNew = {
+        ...childNewThought,
+        contexts: _.uniqBy((childNewThought.contexts || []), "value")
```

Comment:
Since `childNewThought` is guaranteed to have a truthy `contexts` property after `addContext`, you can simplify to: 
```js
contexts: _.uniqBy(childNewThought.contexts, "value")
```